extra/to_movement_ops.py

```diff
@@ -0,0 +1,107 @@
+  real_offset = offset + (sum(x*st for (x,_),st in zip(view.mask, view.strides)) if view.mask else 0)
+  real_real_shape = [s for s,st in zip(real_shape, view.strides) if st]
+  strides = [abs(st) for st in view.strides if st and not isinstance(st, Node)]
+  def sort_by_strides(shape, strides): return sorted(zip(shape, strides), key=lambda k: (k[1],-k[0]), reverse=True), sorted(range(len(strides)), key=lambda k: (strides[k],-real_real_shape[k]), reverse=True)
+  ordered_shape_strides, _ = sort_by_strides(real_real_shape, strides)
+  ordered_shape_strides = [list(s) for s in ordered_shape_strides]
+  if strides:
+    i = 0
+    while i < len(ordered_shape_strides):
+      if i<len(ordered_shape_strides)-1 and ordered_shape_strides[i][1] == ordered_shape_strides[i+1][0]*ordered_shape_strides[i+1][1]:
+        ordered_shape_strides[i+1][0] = ordered_shape_strides[i][0]*ordered_shape_strides[i+1][0]
+      else: i += 1
+    flat_shape = [shape_stride[0] for shape_stride in ordered_shape_strides]
+    flat_strides = [shape_stride[1] for shape_stride in ordered_shape_strides]
+    return (flat_shape, flat_strides, real_offset)
+  return (real_shape, view.strides, real_offset)
+
+def views_equivalent(v1: View, v2: View) -> bool:
+  if v1 == v2: return True
+  return flatten_view(v1) == flatten_view(v2)
+
+
+def st_equivalent(st: ShapeTracker, st_rebuilt: ShapeTracker):
+  views = list(st.views)
+  rebuilt_views = list(st_rebuilt.views)
+  i = 0
+  while i < len(views):
+    view, rebuilt_view = views[i], rebuilt_views[i]
+    if view == rebuilt_view:
+      i += 1
+      continue
+    elif view.shape == rebuilt_view.shape:
+      i += 1
+    # hack to skip expands for overlapped strides
+    else:
+      rebuilt_views.pop(i)
+  return True
+
+def test_rebuild(st: ShapeTracker):
+  rebuilt_st = ShapeTracker.from_shape((get_buffer_size(st.views[0].shape, st.views[0].strides, st.views[0].offset, st.views[0].mask),))
+  for mop, arg in st.to_movement_ops():
+    match mop:
+      case MovementOps.RESHAPE:
+        # shapetracker doesn't allow flattening with -1 but required for MovementOps.RESHAPE
+        if arg == (-1,):
+          rebuilt_st = rebuilt_st.reshape((prod(rebuilt_st.views[-1].shape),))
+        else:
+          rebuilt_st = rebuilt_st.reshape(arg)
+      case MovementOps.PERMUTE:
+        rebuilt_st = rebuilt_st.permute(arg)
+      case MovementOps.EXPAND:
+        if len(arg) != len(rebuilt_st.shape):
+          rebuilt_st = rebuilt_st.reshape((1,*rebuilt_st.shape))
+        rebuilt_st = rebuilt_st.expand(arg)
+      case MovementOps.PAD:
+        rebuilt_st = rebuilt_st.pad(arg)
+      case MovementOps.SHRINK:
+        rebuilt_st = rebuilt_st.shrink(arg)
+      case MovementOps.STRIDE:
+        rebuilt_st = rebuilt_st.stride(arg)
+      case _:
+        raise Exception("invalid mop")
+  rebuilt_st = rebuilt_st.simplify()
+  if len(st.views) != len(rebuilt_st.views):
+    if not set(st.views).issubset(set(rebuilt_st.views)):
```

Comment:
`if a and b`
Also: `asserts behind if statements mean that the if statement will not be optimized away`